### PR TITLE
Add a boolean to inform the architecture of the runing operating system

### LINKED ISSUE #16330: Runtime: let NVDA expose which runtime (32-bit or 64-bit) it is running on
Hi,

Stems from #16304:

### Introduction
As of 2024, Windows 10 is the last version to support 32-bit processors from Intel, AMD, and ARM, with Widows 11 supporintg 64-bit platforms only (AMD64 and ARM64). In 2025, consumer support for Windows 10 will end, with end of support for long-term servicing channel builds planned for 2032 (Windows 10 IoT Enterprise 2021 LTSC).

Originally, NVDA was written for 32-bit Python runtime. With impending end of support for 32-bit platforms, moving to 64-bit became a goal for NV Access and the NVDA community. According to #16304, a phased transition plan is proposed, with full transition tentatively scheduled for 2026.

### Is your feature request related to a problem? Please describe.
During 32-bit to 64-bit transition, it becomes important to know which Python runtime NVDA is running on. This way NVDA Core and add-ons can execute appropriate code paths. For example, when accessing Widows Registry, WoW64 paths must be used if running 32-bit NVDA on 64-bit Windows and vice versa.

At the moment the following can be used to detect 32-bit versus 64-bit Python runtime:

* os.environ (PROCESSOR_ARCHITECTURE variable)
* sys.maxsize > 0x7fffffff (64-bit if greater, 32-bit otherwise)
* platform.architecture()

### Describe the solution you'd like
Expose a flag informing NVDA of 32-bit versus 64-bit platform. This can be either a boolean (True on 64-bit, False on 32-bit) or a string denoting the platform in use (can be set during NVDA compilation).

### Describe alternatives you've considered
Standardize on one or more of the available platform detection options.

### Additional context
Creating a flag informing NVDA Core and add-ons about Python runtime in use communicates willingness by NVDA community to take 64-bit transition seriously and can help people with editing code to take advantage of platform differences.

Thanks.

### PR DESCRIPTION
**Pull request by FelipeZanabria**
### Link to issue number:
Closes #16330
### Summary of the issue:
As of 2024, Windows 10 is the last version to support 32-bit processors from Intel, AMD, and ARM, with Windows 11 supporintg 64-bit platforms only (AMD64 and ARM64). In 2025, consumer support for Windows 10 will end, with end of support for long-term servicing channel builds planned for 2032 (Windows 10 IoT Enterprise 2021 LTSC).
Originally, NVDA was written for 32-bit Python runtime. With impending end of support for 32-bit platforms, moving to 64-bit became a goal for NV Access and the NVDA community. According to #16304, a phased transition plan is proposed, with full transition tentatively scheduled for 2026.
### Description of user facing changes
This way NVDA Core and add-ons can execute appropriate code paths. For example, when accessing Windows Registry, WoW64 paths must be used if running 32-bit NVDA on 64-bit Windows and vice versa.
### Description of development approach
Check runtime using
`winVersion.getWinVer().processorArchitecture.endswith(“64”)`
### Testing strategy:
run Python console and try the next code:
```
from winAPI.constants import IS_64_BIT
print(IS_64_BIT)
```
### Known issues with pull request:
None
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Introduced a feature to check if the system is 64-bit, enhancing compatibility and performance checks.

### CODE REVIEW

**Review by seanbudd — APPROVED** [review]

**Comment by seanbudd** [comment]
@josephsl - this doesn't seem to follow the approach you suggested in #16330, can you comment if this closes the issue as expected?

**Comment by josephsl** [comment]
Hi,

At least the approach checks if 64-bit OS is in use, but not whether NVDA itself is running on native (32/32) or compatibility (32/64) layer. NVDA and add-ons can determine operating system architecture, but not whether NVDA itself is on the same bitness as the operating system, and that's where I went with the original issue (let me know if you'd like me to clarify this in the original issue).

There are also quite a few things that could see some changes, notably copyright header and code comment/code placement.

Thanks.

File: source/winAPI/constants.py [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+# Copyright (C) 2024 NV Access Limited
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+import winVersion
 import enum
 
 
+IS_64_BIT: bool = winVersion.getWinVer().processorArchitecture.endswith("64")
+"""True if the system is 64bit"""
```
Comment (seanbudd) [inline source/winAPI/constants.py:11]: I think we need two flags here. Thoughts @josephsl ?

```suggestion
IS_64_BIT_WINDOWS: bool = winVersion.getWinVer().processorArchitecture.endswith("64")
"""True if the Windows copy is 64bit"""

IS_64_BIT_ARCH: bool = os.environ.get("PROCESSOR_ARCHITEW6432") == "AMD64"
"""True if the system architecture is 64bit"""
```

**Comment by josephsl** [comment]
Hi, I think the two flags should rather be: Windows architecture is 64-bit or not (because in many cases system architecture is hardware architecture), and which architecture NVDA itself is running as (that will be “PROCESSOR_ARCHITECTURE” environment variable). Thanks.

File: source/winAPI/constants.py (hunk 1 above)
Comment (seanbudd) [reply source/winAPI/constants.py:11]: @josephsl perhaps 3?

```suggestion
IS_64_BIT_WINDOWS: bool = winVersion.getWinVer().processorArchitecture.endswith("64")
"""True if the Windows copy is 64bit"""

IS_64_BIT_ARCH: bool = os.environ.get("PROCESSOR_ARCHITEW6432") == "AMD64"
"""True if the system architecture is 64bit"""

IS_64_BIT_NVDA = os.environ.get("PROCESSOR_ARCHITECTURE") == "AMD64"
```

**Comment by josephsl** [comment]
Hi,

I think two flags would suffice - for the most part, 64-bit Windows is equivalent to 64-bit system architecture. Also, the second flag has two criticla issues: 32-bit Windows releases do not have PROCESSOR_ARCHITEW6432, and we are only looking at x64 instead of both AMD64 and ARM64.

Thanks.

File: source/winAPI/constants.py (hunk 1 above)
Comment (seanbudd) [reply source/winAPI/constants.py:11]:
```suggestion
IS_64_BIT_WINDOWS: bool = winVersion.getWinVer().processorArchitecture.endswith("64")
"""True if the Windows copy is 64bit"""

IS_64_BIT_NVDA = os.environ.get("PROCESSOR_ARCHITECTURE") == "AMD64"
```

**Review by josephsl — APPROVED** [review]
I think it would be best to add a changelog for this unless people say it is not needed. The just introduced flags may not be needed once NVDA moves fully to 64-bit architecture. Thanks.

**Comment by seanbudd** [comment]
@josephsl I don't think a changelog entry is useful here - when we are deeper in the migration steps we can provide more useful and thorough information to add-on authors

**Pull request merged**